docs/adr/0002-automating-azure-devops-management-tasks.md

```diff
@@ -0,0 +1,76 @@
+
+## Status
+
+Proposed
+
+## Context
+
+Using a [GitOps][1] approach for managing routine configuration updates for Azure DevOps offers the possibility of:
+* Devolving responsibility for such changes to those with the requirements
+* Scaling-out the management overhead of making such changes
+* Maintaining the required oversight via the pull request mechanism
+* Strong audit trail of configuration changes
+
+Here are some example scenarios:
+1. Managing access to projects
+1. Managing service connections and their security requirements
+
+The premise of these [CodeOps][2] solutions relies on using a trusted Azure Active Directory (AAD) service principal to apply configuration changes instead of users running scripts from their computer or using UI tools.
+
+* Users have access to the git repository and can request changes via pull request
+* Administrators have approval oversight of those pull requests
+* Once approved & merged an automated process runs to apply the updated configuration state from the `main` branch
+* Admin access to the automated process itself is tightly controlled and the credential for the trusted identity should never be known outside of the initial registration process of the automation platform (e.g. registering the service connection with Azure DevOps)
+* The repository history provides an audit trail of configuration changes
+
+This approach has already been used in several scenarios, but never where the configuration being managed related to an Azure DevOps instance.
+
+Currently it is not possible to grant an AAD service principal access to Azure DevOps - [documentation-reference][2].
+
+This blocks the current approach, that would otherwise use an Azure DevOps pipeline to execute the CodeOps process.
+
+The remainder of this ADR outlines two approaches for resolving this issue:
+* Creating a 'service account' user in AAD that the automated process uses to authenticate to Azure DevOps
+* Use AAD delegation to allow an AAD application to impersonate a calling user
+
+### AAD Delegation
+This fundamentally changes the approach and would necessitate a move towards the CodeOps process running as some kind of application that can be interacted with by other parties that have an AAD user context.
+
+This would seem to clash with the goal of not requiring users to have the permissions needed to run the process.
+
+
+### Service Account
+Whilst it would be trivial to setup a dedicated AAD user account with the required permissions to run the process, such an identity cannot always be used to natively execute a pipeline within CI/CD tools (e.g. Azure DevOps, GitHub Actions).
+
+Therefore it would be necessary for the pipeline to perform a 'runas' operation when communicating with the Azure DevOps REST API.  This in turn means that the credential for this user needs to be retrievable (e.g. from key vault), but this clashes with the goal to minimise the surface area for attacking the trusted identity's credential.
```

Comment:
How does this differ from the the need to retrieve the credentials for the service principal that the pipeline operates as?

Reply:
The credentials for the service principal only need to be known at the point that the service connection is registered, after that, ADO handles storing the credentials and making them available to pipeline.

Those concerns, as they relate to this additional identity, would get pushed down into the pipeline process itself (e.g. the codeops script).

Reply:
Could you create a (generic service connection)[https://docs.microsoft.com/en-us/azure/devops/pipelines/library/service-endpoints?view=azure-devops&tabs=yaml#sep-generic] for the service account and handle it in a similar way?